bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/builders/AutomationChannelBuilder.java

```diff
@@ -0,0 +1,244 @@
+/**
+ * Copyright (c) 2010-2022 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.hdpowerview.internal.builders;
+
+import java.time.DayOfWeek;
+import java.time.LocalTime;
+import java.time.format.TextStyle;
+import java.util.ArrayList;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.StringJoiner;
+import java.util.stream.Collectors;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.hdpowerview.internal.HDPowerViewBindingConstants;
+import org.openhab.binding.hdpowerview.internal.HDPowerViewTranslationProvider;
+import org.openhab.binding.hdpowerview.internal.api.responses.SceneCollections.SceneCollection;
+import org.openhab.binding.hdpowerview.internal.api.responses.Scenes.Scene;
+import org.openhab.binding.hdpowerview.internal.api.responses.ScheduledEvents;
+import org.openhab.binding.hdpowerview.internal.api.responses.ScheduledEvents.ScheduledEvent;
+import org.openhab.core.library.CoreItemFactory;
+import org.openhab.core.thing.Channel;
+import org.openhab.core.thing.ChannelGroupUID;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.binding.builder.ChannelBuilder;
+import org.openhab.core.thing.type.ChannelTypeUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link AutomationChannelBuilder} class creates automation channels
+ * from structured scheduled event data.
+ *
+ * @author Jacob Laursen - Initial contribution
+ */
+@NonNullByDefault
+public class AutomationChannelBuilder {
+
+    private final Logger logger = LoggerFactory.getLogger(AutomationChannelBuilder.class);
+    private final HDPowerViewTranslationProvider translationProvider;
+    private final ChannelGroupUID channelGroupUid;
+    private final ChannelTypeUID channelTypeUid = new ChannelTypeUID(HDPowerViewBindingConstants.BINDING_ID,
+            HDPowerViewBindingConstants.CHANNELTYPE_AUTOMATION_ENABLED);
+
+    private List<Channel> channels;
+    private Map<Integer, Scene> scenes;
+    private Map<Integer, SceneCollection> sceneCollections;
+    private List<ScheduledEvent> scheduledEvents;
+
+    public AutomationChannelBuilder(HDPowerViewTranslationProvider translationProvider,
+            ChannelGroupUID channelGroupUid) {
+        this.translationProvider = translationProvider;
+        this.channelGroupUid = channelGroupUid;
+        this.channels = new ArrayList<>(0);
+        this.scheduledEvents = new ArrayList<>(0);
+        this.scenes = new HashMap<>(0);
+        this.sceneCollections = new HashMap<>(0);
```

Comment:
I think you could declare these fields as 'final' rather than initializing them in the constructor; also I think the `(0)` is superfluous..

```
private final List<Channel> channels = new ArrayList<>();
private final Map<Integer, Scene> scenes = new HashMap<>();
private final Map<Integer, SceneCollection> sceneCollections= new HashMap<>();
private final List<ScheduledEvent> scheduledEvents  = new ArrayList<>();
```

Reply:
They are defaults until `withChannels`, `withScenes `etc. will override them, so they can't be final. In the most likely use-case they won't be used, so size 0 seems more appropriate than default allocation (i.e. 10) to save memory. Only channels might actually be extended.

Reply:
In that case you could initialize them to null. Or??

Reply:
Yes, you are right. And now, when doing so, I remember why I stopped and forgot to get back to it the last time - null annotation hell. Fixed.